extra/snake-game/snake-game.factor

```diff
@@ -0,0 +1,395 @@
+
+: offset ( loc dim -- loc )
+    [ + ] 2map ;
+
+: draw-snake-head ( loc facing-dir -- )
+    dup name>> rest "head-" prepend
+    [
+        [ screen-loc ] dip
+        {
+            { :right [ { -20 -10 } ] }
+            { :down  [ { -10 -20 } ] }
+            { :up    [ { -10  0  } ] }
+            { :left  [ {  0  -10 } ] }
+        } case offset
+    ] dip
+    swap draw-sprite* ;
+
+: draw-snake-body ( loc from-dir to-dir -- )
+    2array [ name>> rest ] map "body" prefix "-" join
+    draw-sprite ;
+
+: draw-snake-tail ( loc facing-dir -- )
+    name>> rest "tail-" prepend draw-sprite ;
+
+: draw-snake-part ( loc from-dir snake-part -- )
+    dup type>> {
+        { :head [ drop opposite-dir draw-snake-head ] }
+        { :tail [ drop draw-snake-tail ] }
+        { :body [ dir>> draw-snake-body ] }
+    } case ;
+
+: ?roll-over ( x max -- x )
+    {
+        { [ 2dup >= ] [ 2drop 0 ] }
+        { [ over neg? ] [ nip 1 - ] }
+        [ drop ]
+    } cond ;
+
+: ?roll-over-x ( x -- x )
+    snake-game-dim first ?roll-over ;
+
+: ?roll-over-y ( y -- y )
+    snake-game-dim second ?roll-over ;
+
+: move ( loc dim -- loc )
+    offset first2
+    [ ?roll-over-x ] [ ?roll-over-y ] bi* 2array ;
+
+: relative-loc ( loc dir -- loc )
+    {
+        { :left  [ { -1  0 } move ] }
+        { :right [ {  1  0 } move ] }
+        { :up    [ {  0 -1 } move ] }
+        { :down  [ {  0  1 } move ] }
+    } case ;
+
+: draw-snake-reduce-step ( loc from-dir snake-part -- {new-loc,new-from-dir} )
+    nip dir>> [ relative-loc ] keep 2array ;
+
+: draw-snake ( snake loc from-dir -- )
+    2array 2dup
+    [
+        [ first2 ] dip
+        [ draw-snake-part ] [ draw-snake-reduce-step ] 3bi
+    ] reduce drop
+    ! make sure to draw the head again
+    swap first [ first2 ] dip draw-snake-part ;
+
+: grow-snake ( snake dir -- snake )
+    opposite-dir :head <snake-part> prefix
+    dup second :body >>type drop ;
+
+: snake-shape ( snake -- dirs )
+    [ dir>> ] map ;
+
+: move-snake ( snake dir -- snake )
+    dupd [ snake-shape but-last ] dip
+    opposite-dir prefix [ >>dir ] 2map ;
+
+: update-snake-shape ( snake-game dir growing? -- )
+    [ [ grow-snake ] curry change-snake ]
+    [ [ move-snake ] curry change-snake ]
+    if drop ;
+
+: update-snake-loc ( snake-game dir -- )
+    [ relative-loc ] curry change-snake-loc drop ;
+
+: update-snake-dir ( snake-game dir -- )
+    >>snake-dir drop ;
+
+: point>index ( loc -- n )
+    first2 [ ] [ snake-game-dim first * ] bi* + ;
+
+: index>point ( n -- loc )
+    snake-game-dim first /mod swap 2array ;
+
+: snake-occupied-locs ( snake head-loc -- points )
+    [ dir>> relative-loc ] accumulate nip ;
+
+: snake-occupied-indices ( snake head-loc -- points )
+    snake-occupied-locs [ point>index ] map natural-sort ;
+
+: all-indices ( -- points )
+    snake-game-dim first2 * iota ;
+
+: snake-unoccupied-indices ( snake head-loc -- points )
+    [ all-indices ] 2dip snake-occupied-indices >hash-set without ;
```

Comment:
You don't need ``>hash-set`` here, since ``without`` calls ``tester`` which calls ``fast-set`` which (for sequences) calls ``>hash-set``.